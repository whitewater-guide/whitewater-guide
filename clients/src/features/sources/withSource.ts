import { branch, compose, withProps } from 'recompose';
import { HarvestMode, Source, SourceInput } from '../../../ww-commons';
import { enhancedQuery } from '../../apollo';
import { withFeatureIds } from '../../core';
import { ScriptsList, withScriptsList } from '../scripts';
import REGION_DETAILS from './sourceDetails.query';

export const NEW_SOURCE: SourceInput = {
  id: null,
  name: '',
  termsOfUse: null,
  script: '',
  cron: null,
  harvestMode: HarvestMode.ONE_BY_ONE,
  enabled: false,
  url: null,
};

export interface WithSourceOptions {
  errorOnMissingId?: boolean;
  includeScripts?: boolean;
}

export interface WithSourceChildProps {
  scripts?: ScriptsList;
  source: {
    data: Source | null;
    loading: boolean;
  };
  errors: {
    [key: string]: {[key: string]: any};
  };
}

export interface WithSourceResult {
  source: Source | null;
}

export interface WithSourceProps {
  sourceId?: string;
  language?: string;
}

/**
 *
 * @param options.errorOnMissingId (true) = Should error be added if sourceId is not found?
 * @param options.includeScripts (false) = Should also request scripts?
 * @returns High-order component
 */
export function withSource(options: WithSourceOptions = {}) {
  const { errorOnMissingId = true, includeScripts } = options;
  return compose<WithSourceChildProps, any>(
    withFeatureIds('source'),
    includeScripts ? withScriptsList : ((i: any) => i),
    // If no source was found, branch provides dummy source with error
    branch<WithSourceProps>(
      ({ sourceId }) => !!sourceId,
      enhancedQuery<WithSourceResult, WithSourceProps, WithSourceChildProps>(
        REGION_DETAILS,
        {
          options: ({ sourceId, language }) => ({
            fetchPolicy: 'cache-and-network',
            variables: { id: sourceId, language },
          }),
          props: ({ data }) => {
            const { source, loading } = data!;
            return {
              source: {
                data: source,
                loading: loading && !source,
              },
            };
          },
          alias: 'withSource',
        },
      ),
      withProps<WithSourceChildProps, WithSourceProps>(({ sourceId }) => (
        {
          source: {
            data: NEW_SOURCE as any,
            loading: false,
          },
          errors: errorOnMissingId ? {
            sourceDetails: { message: `No source with id ${sourceId} was found` },
          } : {},
        }
      )),
    ),
  );
}

export type WithSource = WithSourceProps & WithSourceChildProps;
