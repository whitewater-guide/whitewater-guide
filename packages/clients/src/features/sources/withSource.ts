import { FetchPolicy } from 'apollo-client';
import { graphql } from 'react-apollo';
import { compose } from 'recompose';
import { Source } from '../../../ww-commons';
import { queryResultToNode, WithNode } from '../../apollo';
import { withFeatureIds } from '../../core';
import SOURCE_DETAILS from './sourceDetails.query';

interface WithSourceOptions {
  fetchPolicy?: FetchPolicy;
}

interface WithSourceResult {
  source: Source;
  sourceId: string;
}

interface WithSourceProps {
  sourceId: string;
}

export interface WithSource {
  source: WithNode<Source>;
  sourceId: string;
}

export const withSource = ({ fetchPolicy = 'cache-and-network' }: WithSourceOptions = {}) =>
  compose<WithSource, any>(
    withFeatureIds('source'),
    graphql<WithSourceResult, WithSourceProps, WithSource>(
      SOURCE_DETAILS,
      {
        alias: 'withSource',
        options: { fetchPolicy },
        props: props => queryResultToNode(props, 'source'),
      },
    ),
  );
