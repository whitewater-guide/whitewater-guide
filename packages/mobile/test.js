const MarkdownIt = require('markdown-it');
const md = new MarkdownIt();
const res = MarkdownIt.parseInline(
  'The content you are adding will be available under an open licence (CC). See our [Terms of Use](1) and others',
);

const nodes = [];
let currentLink;
res[0].children.forEach((child) => {
  if (child.type === 'text') {
    if (currentLink) {
      currentLink.text = child.content;
    } else {
      nodes.push(child.content);
    }
  } else if (child.type === 'link_open') {
    currentLink = { href: child.attrGet('href') };
  } else if (child.type === 'link_close') {
    nodes.push(currentLink);
    currentLink = null;
  }
});
console.log(nodes);
