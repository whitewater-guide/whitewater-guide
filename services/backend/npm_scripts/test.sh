#!/usr/bin/env bash

set -o allexport
source ../../config/.env.test
source .env.test
set +o allexport

./node_modules/.bin/jest -i --forceExit "$@"
