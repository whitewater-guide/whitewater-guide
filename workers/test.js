const { execFile } = require('child_process');

execFile('./bin/norway', ['harvest', '--code=213.4', '--html'], (error, stdout, stderr) => {
  if (error) {
    throw error;
  }
  const result = JSON.parse(stdout)
  console.log(JSON.stringify(result, null, 2))
  // console.log(result.success, result.data.length)
});



