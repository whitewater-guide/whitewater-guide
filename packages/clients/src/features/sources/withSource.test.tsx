import { mount } from 'enzyme';
import React from 'react';
import { createMockedProvider, flushPromises, Receiver } from '../../test';
import { WithSource, withSource } from './withSource';

let receiver: Receiver<WithSource>;

beforeEach(async () => {
  const MockedProvider = createMockedProvider();
  const WithData = withSource()(Receiver as any);
  const wrapped = mount(
    <MockedProvider>
      <WithData sourceId="foo" />
    </MockedProvider>,
  );
  receiver = wrapped
    .find(Receiver as any)
    .first()
    .instance() as any;
  await flushPromises();
});

it('should have loading state', async () => {
  expect(receiver.cwrp[0].source.loading).toBe(true);
  expect(receiver.cwrp[1].source.loading).toBe(false);
});

it('should match snapshot', () => {
  expect(receiver.cwrp[1].source).toMatchSnapshot();
});
