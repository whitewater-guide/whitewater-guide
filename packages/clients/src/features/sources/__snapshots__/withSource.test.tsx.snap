// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
Object {
  "error": undefined,
  "loading": false,
  "networkStatus": 7,
  "node": Object {
    "__typename": "Source",
    "cron": "Source.cron.1",
    "enabled": false,
    "harvestMode": "allAtOnce",
    "id": "Source.id.1",
    "name": "Source.name.1",
    "regions": Object {
      "__typename": "SourceRegionConnection",
      "nodes": Array [
        Object {
          "__typename": "Region",
          "id": "Region.id.1",
          "name": "Region.name.1",
        },
        Object {
          "__typename": "Region",
          "id": "Region.id.2",
          "name": "Region.name.2",
        },
      ],
    },
    "script": "Source.script.1",
    "termsOfUse": "Source.termsOfUse.1",
    "url": "Source.url.1",
  },
  "refetch": [Function],
}
`;
